File: Python/Text Justification.py
class Solution:
    # @param words, a list of strings
    # @param L, an integer
    # @return a list of strings
    def fullJustify(self, words, L):
		res = []
		i = 0
		while i<len(words):
			size=0
			begin=i
			while i < len(words):
				newsize = len(words[i]) if size==0 else size+len(words[i])+1#count in at least one space between words
				if newsize<=L:size = newsize
				else:break
				i = i + 1
			spaceCount = L-size
			if i-begin-1>0 and i<len(words): #i-begin-1 is 0 or bigger than 0: means how many words in this line. not last line and more than one word
				everyCount = spaceCount/(i-begin-1)
				spaceCount = spaceCount%(i-begin-1)
			else:
				everyCount=0
			j=begin
			while j<i:
				if j==begin:s=words[j]
				else:
					s = s + " "*(everyCount+1)
					if spaceCount>0 and i<len(words):
						s = s + " "
						spaceCount = spaceCount - 1
					s = s + words[j]
				j = j + 1
			s = s + " "*spaceCount
			res.append(s)
		return res

File: Java/Text Justification.java
class Text Justification{
	public List<String> fullJustify(String[] words, int L) {
        ArrayList<String> res = new ArrayList<String>();
        int curr = 0;
        if(words == null || words.length == 0)
        	return res;
        while(curr < words.length){
        	int spaceCount = 0;
        	int size = 0;
        	int newSize = 0;
        	int start = curr;
        	while(curr < words.length){
        		newSize = size == 0?words[curr].length():size+words[curr].length()+1;
        		if(newSize > L){
        			break;
        		}
        		else{
        			size = newSize;
        			curr++;
        		}
        	}
        	spaceCount = L - size;
        	int interSpaceCount = 0;
        	int remainSpaceCount = 0;
        	if(curr - start  > 1 && curr < words.length){
        		interSpaceCount = spaceCount/(curr-start-1);
        		remainSpaceCount = spaceCount%(curr-start-1);
        	}
        	else
        		remainSpaceCount = spaceCount;
        	int i = start;
        	String temp = "";
        	while(i<curr){
        		if(i == start){
        			temp = words[i];
        			i++;
        		}
        		else{
        			temp = temp + new String(new char[interSpaceCount+1]).replace('\0',' ');
        			if(remainSpaceCount > 0 && curr<words.length){
        				temp = temp + " ";
        				remainSpaceCount--;
        			}
        			temp = temp + words[i];
        			i++;
        		}
        	}
        	temp = temp + new String(new char[remainSpaceCount]).replace('\0',' ');
        	res.add(temp);
        }
        return res;
    }
}